pixel_clusterizer/cluster_functions.py

```diff
@@ -1,182 +1,277 @@
 ''' Fast clustering functions that are compiled in time via numba '''
-import numpy as np
 from numba import njit
 from docutils.utils.roman import OutOfRangeError
 
 
 @njit()
-def _new_event(event_number, actual_event_number):
+def _new_event(event_number_1, event_number_2):
     'Detect a new event by checking if the event number of the actual hit is the actual event number'
-    return event_number != actual_event_number
+    return event_number_1 != event_number_2
 
 
 @njit()
-def _finish_event(hits, cluster, is_seed, n_cluster, cluster_size, cluster_id, actual_event_hit_index, new_actual_event_hit_index, next_cluster_id, actual_event_cluster_index):
-    ''' Set hit and cluster information of the last finished event (like number of cluster in this event (n_cluster),  cluster charge ...). '''
-    for i in range(actual_event_hit_index, new_actual_event_hit_index):  # Set hit cluster info that is only known at the end of the event
-        n_cluster[i] = next_cluster_id
+def _pixel_masked(hit, array):
+    ''' Checks whether a hit (column/row) is masked or not. Array is 2D array with boolean elements corresponding to pixles indicating whether a pixel is disabled or not.
+    '''
+    if array.shape[0] > hit["column"] and array.shape[1] > hit["row"]:
+        return array[hit["column"], hit["row"]]
+    else:
+        return False
 
-    # Normalize cluster index by the charge for center of gravity
-    for i in range(actual_event_cluster_index, actual_event_cluster_index + next_cluster_id):
-        cluster[i]['mean_column'] /= (cluster[i]['charge'] + cluster[i]['n_hits'])
-        cluster[i]['mean_row'] /= (cluster[i]['charge'] + cluster[i]['n_hits'])
+
+# @njit()
+# def _pixel_masked(hit, array):
+#     ''' Checks whether a hit (column/row) is masked or not. Array is an iterable of column/row tuples of disabled pixels.
+#     '''
+#     for array_value in array:
+#         if hit["column"] == array_value["column"] and hit["row"] == array_value["row"]:
+#             return True
+#     return False
+
+
+@njit()
+def _finish_cluster(hits, cluster, cluster_size, cluster_hit_indices, cluster_index, cluster_id, charge_correction, noisy_pixels, disabled_pixels):
+    ''' Set hit and cluster information of the cluster (e.g. number of hits in the cluster (cluster_size), total cluster charge (charge), ...).
+    '''
+    cluster_charge = 0
+    max_cluster_charge = -1
+    # necessary for charge weighted hit position
+    total_weighted_column = 0
+    total_weighted_row = 0
+
+    for i in range(cluster_size):
```

Comment:
If I read this correctly it means that all clusters are looped at the end of each cluster. That is a `factorial(n_cluster)` run time algorithm, before it was linear.